providers/dns/utils.go

```diff
@@ -0,0 +1,47 @@
+package dns
+
+import (
+	"fmt"
+	"time"
+)
+
+// ToFqdn converts the name into a fqdn appending a trailing dot.
+func ToFqdn(name string) string {
+	n := len(name)
+	if n == 0 || name[n-1] == '.' {
+		return name
+	}
+	return name + "."
+}
+
+// UnFqdn converts the fqdn into a name removing the trailing dot.
+func UnFqdn(name string) string {
+	n := len(name)
+	if n != 0 && name[n-1] == '.' {
+		return name[:n-1]
+	}
+	return name
+}
+
+// WaitFor polls the given function 'f', once every 'interval' seconds, up to 'timeout' seconds.
+func WaitFor(timeout, interval int, f func() (bool, error)) error {
```

Comment:
This function is now duplicated, declared in both `providers/dns/utils.go` and `acme/dns_challenge.go`. Since it is not specific to DNS providers i suggest to declare this once in `acme/utils.go`.